feat: allow installation with `requirements.txt`

Luke-zhang-04 opened the pull request:
- merge branch upstream with master 33079ad
- docs: update docstirngs to fit with PEP conventions 42ce232
- feat: add `requirements.txt` install function 538d6f2
- feat: users can install with requirements.txt efa4936
- refactor: format code with psf/black 6dd89a0
- docs: add instructions for requirements.txt installation 4fba1c1
- move release to one file, test with and without requirements 668608b
- fix(install_packages): check for array length 325c486
- update examples 6dab958 
- feat: users can install with requirements.txt efa4936
- refactor: format code with psf/black 6dd89a0
- docs: add instructions for requirements.txt installation 4fba1c1
- move release to one file, test with and without requirements 668608b
- fix(install_packages): check for array length 325c486
- update examples 6dab958 
-  fix(CI): fix workflow syntax error on line 36 c55fd36
- refactor: remove f-strings effadfc 
	- f-strings incompatible with Python versions below 3.6, which is the target audience here
-  add pull_request to workflow checking 9015060

Luke-zhang-04 committed:
simplified imports

Luke-zhang-04 committed:
added __init__.py

Luke-zhang-04 committed:
added readme

Luke-zhang-04 committed:
added license

Luke-zhang-04 committed:
updated readme

Luke-zhang-04 committed:
updated readme

Luke-zhang-04 committed:
added gitignore

Luke-zhang-04 committed:
renamed lib to pypackageinstaller

Luke-zhang-04 committed:
updated readme

Luke-zhang-04 committed:
fixed project example and changed lib file

Luke-zhang-04 committed:
added underscore to pil examples, fixed examples with error file not found

Luke-zhang-04 committed:
updated readme

Luke-zhang-04 committed:
fixed readme

Luke-zhang-04 committed:
reverted example changes

Luke-zhang-04 committed:
changed pypackageinstaller -> pipinstaller

Luke-zhang-04 committed:
moved resources into pipinstaller

Luke-zhang-04 committed:
update license

Luke-zhang-04 committed:
fixed problem on windows with string escaping

Luke-zhang-04 committed:
merge branch upstream with master

Signed-off-by: Luke-zhang-04 <luke.zhang2004dev@gmail.com>

Luke-zhang-04 committed:
docs: update docstirngs to fit with PEP conventions

Signed-off-by: Luke-zhang-04 <luke.zhang2004dev@gmail.com>

Luke-zhang-04 committed:
feat: add `requirements.txt` install function

Signed-off-by: Luke-zhang-04 <luke.zhang2004dev@gmail.com>

kertox662 commented on examples/ProjectExample/pipinstaller/install_Functions.py:
```diff
@@ -43,9 +41,7 @@ def checkPackage(pkg):
 def installPip(printOut = False):
-    pipPath = resourcesPath / "get-pip.py"
-    print(pipPath)
-    cmd = f"py -3 {pipPath} --user" if sys.platform == "win32" else f"python3 {pipPath} --user"
```
Wait, what was wrong here?

kertox662 replied:
might want to add command though instead. f"{cmd} {pipPath}" if this is still valid

kertox662 commented on pipinstaller/install_Functions.py:
```diff
@@ -15,7 +14,8 @@
 #######################################
 def checkPip():
-    cmd = "pip"
+    cmd = "py -3 -m pip" if sys.platform == 'win32' else "python3 -m pip" #This command will normally give options for 
```
might want to do specific python command instead.
```python
pycmd = "python3"
if sys.platform == 'win32':
    pycmd = "py -3"
cmd = "{cmd} -m pip"
```

kertox662 replied:
easier to modify later

kertox662 commented on pipinstaller/install_Functions.py:
```diff
@@ -71,7 +70,7 @@ def installPip(printOut = False):
 #######################################
 def installPackage(pkg, printOut = False):
-    cmd = f"pip install {pkg[1]} --user"
+    cmd = "py -3 -m pip install {} --user".format(pkg[1]) if sys.platform == 'win32' else "python3 -m pip install {} --user".format(pkg[1])
```
same as above, feel f-string might be cleaner.
```python
cmd = f"{pycmd} -m pip install {pkg[1]} --user"
```
I agree though that it should change

Luke-zhang-04 replied:
F strings are not supported in python 3.5 and below
[https://www.python.org/dev/peps/pep-0498/](https://www.python.org/dev/peps/pep-0498/)
See effadfc

kertox662 reviewed (changes requested):
The other changes also fall into the same thing. I do think using the path module is cleaner than hardcoding the path with the separators as strings, don't know if there's any specific issue with it.

kertox662 commented on examples/ProjectExample/pipinstaller/install_Functions.py:
```diff
@@ -25,76 +23,83 @@ def checkPip():
-    print(pkg[0],"was found on the system")
-    return True #Returns if whether it is installed or not
+    print(pkg[0], "was found on the system")
+    return True # Returns if whether it is installed or not
```
Small thing, could you quickly change it to "Returns whether it ...", just remove the if. I think I had something there before, doesn't make sense anymore.

kertox662 commented:
Just curious if you know why all 20 commits are opened up now instead of just the modified ones. Is it because of commit squashing? I'm pretty sure though it should only be 2 or so if I'm not wrong.

Luke-zhang-04 committed:
feat: users can install with requirements.txt

version specification
automatic checking for installation
easy to use, supported officially

Signed-off-by: Luke-zhang-04 <luke.zhang2004dev@gmail.com>

Luke-zhang-04 committed:
refactor: format code with psf/black

Signed-off-by: Luke-zhang-04 <luke.zhang2004dev@gmail.com>

Luke-zhang-04 committed:
docs: add instructions for requirements.txt installation

Signed-off-by: Luke-zhang-04 <luke.zhang2004dev@gmail.com>

Luke-zhang-04 committed:
move release to one file, test with and without requirements

Signed-off-by: Luke-zhang-04 <luke.zhang2004dev@gmail.com>

Luke-zhang-04 committed:
fix(install_packages): check for array length

array takes priority

Signed-off-by: Luke-zhang-04 <luke.zhang2004dev@gmail.com>

Luke-zhang-04 committed:
update examples

Signed-off-by: Luke-zhang-04 <luke.zhang2004dev@gmail.com>

Luke-zhang-04 commented:
> Just curious if you know why all 20 commits are opened up now instead of just the modified ones. Is it because of commit squashing? I'm pretty sure though it should only be 2 or so if I'm not wrong.

I don't know

Luke-zhang-04 committed:
fix(CI): fix workflow syntax error on line 36

Signed-off-by: Luke-zhang-04 <luke.zhang2004dev@gmail.com>

Luke-zhang-04 committed:
update comment

Signed-off-by: Luke-zhang-04 <luke.zhang2004dev@gmail.com>

Luke-zhang-04 committed:
fix workflow

Signed-off-by: Luke-zhang-04 <luke.zhang2004dev@gmail.com>

Luke-zhang-04 committed:
fix workflow

Signed-off-by: Luke-zhang-04 <luke.zhang2004dev@gmail.com>

Luke-zhang-04 committed:
fix workflow

Signed-off-by: Luke-zhang-04 <luke.zhang2004dev@gmail.com>

Luke-zhang-04 committed:
refactor: remove f-strings

f-strings not compatible with Python versions below 3.6

Signed-off-by: Luke-zhang-04 <luke.zhang2004dev@gmail.com>

Luke-zhang-04 committed:
fix workflow

Signed-off-by: Luke-zhang-04 <luke.zhang2004dev@gmail.com>

Luke-zhang-04 committed:
add pull_request to workflow checking

Signed-off-by: Luke-zhang-04 <luke.zhang2004dev@gmail.com>

Luke-zhang-04 committed:
use pippath and respath from pathlib instead of hardcoded path

Signed-off-by: Luke-zhang-04 <luke.zhang2004dev@gmail.com>

Luke-zhang-04 committed:
use pathlib for requirements.txt

Signed-off-by: Luke-zhang-04 <luke.zhang2004dev@gmail.com>

Luke-zhang-04 committed:
specify Python 3.5 for CI

Signed-off-by: Luke-zhang-04 <luke.zhang2004dev@gmail.com>
